diff --git a/node_modules/@antv/g6/lib/types/element.d.ts b/node_modules/@antv/g6/lib/types/element.d.ts
index 391930d..f9a8772 100644
--- a/node_modules/@antv/g6/lib/types/element.d.ts
+++ b/node_modules/@antv/g6/lib/types/element.d.ts
@@ -7,6 +7,8 @@ import type { Point, Port } from '../types';
  * <en/> Node type
  */
 export interface Node extends DisplayObject, ElementHooks, ElementMethods {
+    type: "node",
+    id: string,
     /**
      * <zh/> 获取连接桩
      *
@@ -39,13 +41,16 @@ export interface Node extends DisplayObject, ElementHooks, ElementMethods {
  * <en/> Edge type
  */
 export interface Edge extends DisplayObject, ElementHooks, ElementMethods {
+    type: "edge",
+    id: string,
 }
 /**
  * <zh/> 组合类型
  *
  * <en/> Combo type
  */
-export interface Combo extends Node {
+export interface Combo extends Omit<Node, "type"> {
+    type: "combo",
     /**
      * <zh/> 获取组合的位置
      *
